### DISCUSSION TITLE
File extension validation

### DISCUSSION THREAD

**Discussion by samjoy** [fastapi/fastapi#6680] [Questions] [answered]
Is there a way to validate the file format of an input file as a request? If the fastapi app will only accept incoming requests with files that have extensions .pdf,.docx or .txt, how will I implement this using fastapi? Something similar is used in the laravel framework https://stackoverflow.com/questions/32179164/how-to-validate-a-file-type-in-laravel

### Example

```Python
from fastapi import FastAPI

app = FastAPI()


@app.post("/upload")
def read_file(input_file: UploadFile = File(..., format=[".txt",".pdf",".docx"])):
    return {"filename": "input_file.filename"}
```

**Comment by frankie567** [comment, marked as answer]
You can check the [MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types), which should be more accurate than the file extension (https://fastapi.tiangolo.com/tutorial/request-files/#uploadfile). Example:

```py
@app.post("/upload")
def read_file(input_file: UploadFile = File(...)):
    if input_file.content_type not in ["application/pdf", "text/plain", "application/vnd.openxmlformats-officedocument.wordprocessingml.document"]:
        raise HTTPException(400, detail="Invalid document type")
    return {"filename": "input_file.filename"}
```

Of course, you can wrap this logic in a dependency to reuse it in several endpoints.

**Comment by samjoy** [comment]
@frankie567 Thank you

**Comment by nachiket-lab** [comment]
Hi @frankie567 

So, I just tried to do this in my upload endpoint.

```python
if file.content_type not in ["application/json", "text/plain", "text/csv"]:
  raise HTTPException(
      status_code=400,
      detail=f"File type of {file.content_type} is not supported",
  )
```

But I am still able to upload binaries or other files by specifying the type in curl. Is  there any way we could stop this?
```bash
curl -X 'POST' 'http://127.0.0.1:8000/opa/create/file/nachiket%2F'   \
  -H 'accept: application/json'   \
  -H 'Content-Type: multipart/form-data'   \
  -F 'file=@/bin/bash;type=application/json'

**Comment by frankie567** [comment]
Just by looking at `content_type`, no, I don't think so. If you want something more robust, you'll have to use a library guessing the type of the file from its content, like [python-magic](https://github.com/ahupp/python-magic).

**Comment by nachiket-lab** [comment]
Yup...! Thought so. The issue would be loading the file into a buffer by using `file.read()` and then using `python-magic`. Larger files may be an issue in that situation, cause it would be necessary to read the file in memory. But I guess it can't be helped.
Or I could write to a temp location and then do a validation, but then the response would be slow.

**Comment by aalvrz** [comment]
@frankie567 What is the correct way to set the file content type in a test using `TestClient`? The following gives me a blank `""` string...:

```python
from io import BytesIO

from fastapi.testclient import TestClient


client = TestClient(app)

response = client.post(
            "foobar/",
            files={
                "floorplan": (
                    "myimage.png",
                    BytesIO(faker.image()),
                )
            },
        )
```

**Comment by tiangolo** [comment]
Thanks for the help @frankie567 ! 🍰

And thanks for reporting back and closing the issue @samjoy ! ☕

@aalvrz the current test client is underneath just the Requests library, so whatever works there would work here.

In a future version (soon) the test client will be based on HTTPX, but the API is almost identical.

**Comment by aalvrz** [comment]
@tiangolo Thanks for the response. It is still not very clear, passing explicit content type header seems to fail (422). Perhaps because passing `files` in the request automatically sets it to be multipart content? Not sure:

```python
@app.post("/upload", status_code=201)
def upload_file(floorplan: UploadFile):
    pass


def test_upload():
    response = test_client.post(
           "/upload",
            headers={"Content-Type": "image/jpeg"},
            files={
                "floorplan": (
                    "myimage.png",
                    BytesIO(faker.image()),
                )
            },
        )
```

Response content:

```python
{'detail': [{'loc': ['body', 'floorplan'], 'msg': 'field required', 'type': 'value_error.missing'}]}
```

**Comment by frankie567** [comment]
Actually, you need to set the content type of the file, not the whole request.

The file tuple parameter can accept a third value which is the content type of the file: https://requests.readthedocs.io/en/latest/user/advanced/#post-multiple-multipart-encoded-files

**Comment by Princekrampah** [comment]
```python3
import filetype
from typing import IO
from fastapi import HTTPException, status


async def validate_file_size_type(file: IO):
    FILE_SIZE = 2097152  # 2MB

    accepted_file_types = [
        "image/png",
        "image/jpeg",
        "image/jpg",
        "png",
        "jpeg",
        "jpg",
    ]

    file_info = filetype.guess(file.file)

    if file_info is None:
        raise HTTPException(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            detail="Unable to determine file type",
        )

    detected_content_type = file_info.extension.lower()

    if (
        file.content_type not in accepted_file_types
        or detected_content_type not in accepted_file_types
    ):
        raise HTTPException(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            detail="Unsupported file type",
        )

    real_file_size = 0
    for chunk in file.file:
        real_file_size += len(chunk)
        if real_file_size > FILE_SIZE:
            raise HTTPException(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, detail="Too large")

    # return the pointer back to the starting point so that the next read starts from the starting point
    await file.seek(0)
```
